crates/pipeline_manager/src/api/services/probe.rs

```diff
@@ -0,0 +1,235 @@
+                KafkaServiceProbeRequest::GetTopics => "kafka_get_topics".to_string(),
+            },
+        }
+    }
+
+    /// Deserialize from provided YAML.
+    pub fn from_yaml_str(s: &str) -> Self {
+        serde_yaml::from_str(s).unwrap()
+    }
+
+    /// Serialize to YAML for storage.
+    pub fn to_yaml(&self) -> String {
+        serde_yaml::to_string(&self).unwrap()
+    }
+}
+
+/// Enumeration of all possible service probe success responses.
+#[derive(Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceProbeSuccess {
+    /// Successful result from a connectivity probe.
+    Connected,
+    /// Successful result from a probe specific to a Kafka service.
+    Kafka(KafkaServiceProbeSuccess),
+}
+
+/// Range of possible errors that can occur during a service probe.
+/// These are shared across all services.
+#[derive(ThisError, Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub enum ServiceProbeError {
+    #[error("Timeout was exceeded")]
+    TimeoutExceeded,
+    #[error("Request type ({request_type:?}) was issued to a service type ({service_type}) which does not support it")]
+    UnsupportedRequest {
+        request_type: String,
+        service_type: String,
+    },
+    #[error("{0}")]
+    Other(String),
+}
+
+/// Response being either success or error.
+#[derive(Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceProbeResponse {
+    // Serialization of the enums are done using with::singleton_map
+    // because serde_yaml does not support nested enums currently.
+    // See also: <https://github.com/dtolnay/serde-yaml/issues/363>
+    #[serde(with = "serde_yaml::with::singleton_map")]
+    Success(ServiceProbeSuccess),
+    #[serde(with = "serde_yaml::with::singleton_map")]
+    Error(ServiceProbeError),
+}
+
+impl ServiceProbeResponse {
+    /// Convert result into response.
+    pub fn from_result(result: ServiceProbeResult) -> Self {
+        match result {
+            Ok(s) => ServiceProbeResponse::Success(s),
+            Err(e) => ServiceProbeResponse::Error(e),
+        }
+    }
+
+    /// Deserialize from provided YAML.
+    pub fn from_yaml_str(s: &str) -> Self {
+        serde_yaml::from_str(s).unwrap()
+    }
+
+    /// Serialize to YAML for storage.
+    pub fn to_yaml(&self) -> String {
+        serde_yaml::to_string(&self).unwrap()
+    }
+}
+
+/// Result of a service probe.
+pub type ServiceProbeResult = Result<ServiceProbeSuccess, ServiceProbeError>;
+
+#[cfg(test)]
+mod tests {
+    use crate::api::{KafkaServiceProbeRequest, ServiceProbeRequest, ServiceProbeStatus};
+    use crate::db::DBError;
+
+    #[test]
+    fn test_status_string_conversion() {
```

Comment:
These tests are a little too paranoid. :) Are you worried about catching inadvertent edits to the From/ToString results?

Reply:
Yes, effectively that what is written into and read from the database is what is expected. I agree that it is on the cautious side, but the test was quick to write and it being a database type, (de-)serialization is very important to remain consistent (as you mentioned, in case of inadvertent edits, or other like addition of a new status).